src/pkg/packager/lint/lint.go

```diff
@@ -25,120 +22,109 @@ import (
-
-	if validator.jsonSchema, err = getSchemaFile(); err != nil {
+	var untypedZarfPackage interface{}
+	if err := utils.ReadYaml(layout.ZarfYAML, &untypedZarfPackage); err != nil {
 		return nil, err
 	}
 
-	if err = validateSchema(&validator); err != nil {
+	schemaFindings, err := validateSchema(jsonSchema, untypedZarfPackage)
+	if err != nil {
 		return nil, err
 	}
+	findings = append(findings, schemaFindings...)
 
-	return &validator, nil
+	return findings, nil
 }
 
-func lintComponents(ctx context.Context, validator *Validator, createOpts *types.ZarfCreateOptions) {
-	for i, component := range validator.typedZarfPackage.Components {
-		arch := config.GetArch(validator.typedZarfPackage.Metadata.Architecture)
+func lintComponents(ctx context.Context, pkg types.ZarfPackage, createOpts types.ZarfCreateOptions) ([]types.PackageFinding, error) {
+	var findings []types.PackageFinding
 
+	for i, component := range pkg.Components {
+		arch := config.GetArch(pkg.Metadata.Architecture)
 		if !composer.CompatibleComponent(component, arch, createOpts.Flavor) {
 			continue
 		}
 
-		chain, err := composer.NewImportChain(ctx, component, i, validator.typedZarfPackage.Metadata.Name, arch, createOpts.Flavor)
-		baseComponent := chain.Head()
+		chain, err := composer.NewImportChain(ctx, component, i, pkg.Metadata.Name, arch, createOpts.Flavor)
 
-		var badImportYqPath string
-		if baseComponent != nil {
-			if baseComponent.Import.URL != "" {
-				badImportYqPath = fmt.Sprintf(".components.[%d].import.url", i)
-			}
-			if baseComponent.Import.Path != "" {
-				badImportYqPath = fmt.Sprintf(".components.[%d].import.path", i)
-			}
-		}
 		if err != nil {
-			validator.addError(validatorMessage{
-				description:    err.Error(),
-				packageRelPath: ".",
-				packageName:    validator.typedZarfPackage.Metadata.Name,
-				yqPath:         badImportYqPath,
-			})
+			return nil, err
 		}
 
-		node := baseComponent
+		node := chain.Head()
 		for node != nil {
-			checkForVarInComponentImport(validator, node)
-			fillComponentTemplate(validator, node, createOpts)
-			lintComponent(validator, node)
+			component := node.ZarfComponent
+			compFindings := fillComponentTemplate(&component, &createOpts)
+			compFindings = append(compFindings, checkComponent(component, node.Index())...)
+			for i := range compFindings {
+				compFindings[i].PackagePathOverride = node.ImportLocation()
+				compFindings[i].PackageNameOverride = node.OriginalPackageName()
+			}
+			findings = append(findings, compFindings...)
 			node = node.Next()
 		}
 	}
+	return findings, nil
 }
 
-func fillComponentTemplate(validator *Validator, node *composer.Node, createOpts *types.ZarfCreateOptions) {
-	err := creator.ReloadComponentTemplate(&node.ZarfComponent)
+func fillComponentTemplate(c *types.ZarfComponent, createOpts *types.ZarfCreateOptions) []types.PackageFinding {
+	err := creator.ReloadComponentTemplate(c)
+	var findings []types.PackageFinding
```

Comment:
Flip this and the previous line to show that we are immediately checking the error.